rl/epstein_civil_violence/model.py

```diff
@@ -0,0 +1,167 @@
+        observation = {}
+        for agent in self.schedule.agents:
+            observation[agent.unique_id] = [
+                self.obs_grid[neighbor[0]][neighbor[1]]
+                for neighbor in agent.neighborhood
+            ]  # Get the values from the observation grid for the neighborhood cells
+
+        # RL specific outputs for the environment
+        done = {a.unique_id: False for a in self.schedule.agents}
+        truncated = {a.unique_id: False for a in self.schedule.agents}
+        truncated["__all__"] = np.all(list(truncated.values()))
+        if self.schedule.time > self.max_iters:
+            done["__all__"] = True
+        else:
+            done["__all__"] = False
+
+        return observation, rewards, done, truncated, {}
+
+    def cal_reward(self):
+        rewards = {}
+        for agent in self.schedule.agents:
+            if isinstance(agent, CopRL):
+                if agent.arrest_made:
+                    # Cop is rewarded for making an arrest
+                    rewards[agent.unique_id] = 1
+                else:
+                    rewards[agent.unique_id] = 0
+            else:
+                # An active agent is rewarded for its grievance
+                # A jailed agent is penalized for its risk aversion
+                if agent.jail_sentence > 0:
+                    rewards[agent.unique_id] = -agent.risk_aversion
+                else:
+                    rewards[agent.unique_id] = (
+                        0 if agent.condition == "Quiescent" else agent.grievance * 3
+                    )
+
+        return rewards
+
+    def reset(self, *, seed=None, options=None):
+        """
+        Reset the environment after each episode.
+
+        Parameters:
+        - seed: Seed for random number generation.
+        - options: Additional options for resetting the environment.
+
+        Returns:
+        - observation: Initial observation of the environment.
+        - info: Additional information about the reset.
+        """
+
+        super().reset()
+        # Using base scheduler to maintain the order of agents
+        self.schedule = mesa.time.BaseScheduler(self)
+        self.grid = mesa.space.SingleGrid(self.width, self.height, torus=True)
+        create_intial_agents(self, CitizenRL, CopRL)
+        grid_to_observation(self, CitizenRL)
+        # Intialize action dictionary with no action
+        self.action_dict = {a.unique_id: (0, 0) for a in self.schedule.agents}
+        # Update neighbors for observation space
+        for agent in self.schedule.agents:
+            agent.update_neighbors()
+        self.schedule.step()
+        observation = {}
+        for agent in self.schedule.agents:
+            observation[agent.unique_id] = [
+                self.obs_grid[neighbor[0]][neighbor[1]]
+                for neighbor in agent.neighborhood
+            ]  # Get the values from the observation grid for the neighborhood cells
+        return observation, {}
```

Comment:
This seems a bit weird. Can't we call init again or something like that?

Reply:
The problem is with init we call Citizen class whereas we want CItizenRL class hence the need to reimplement it